lib/game/game_widget.dart

```diff
@@ -1,100 +1,265 @@
-import 'package:flutter/gestures.dart';
 import 'package:flutter/widgets.dart';
+import 'package:flutter/gestures.dart';
 
-import '../components/mixins/tapable.dart';
-import '../gestures.dart';
-import 'embedded_game_widget.dart';
 import 'game.dart';
+import 'embedded_game_widget.dart';
+import '../gestures.dart';
+import '../components/mixins/tapable.dart';
 
-bool _hasBasicGestureDetectors(Game game) =>
-    game is TapDetector ||
-    game is SecondaryTapDetector ||
-    game is DoubleTapDetector ||
-    game is LongPressDetector ||
-    game is VerticalDragDetector ||
-    game is HorizontalDragDetector ||
-    game is ForcePressDetector ||
-    game is PanDetector ||
-    game is ScaleDetector;
+typedef GameLoadingWidgetBuilder = Widget Function(
+  BuildContext,
+  bool error,
+);
+
+/// A [StatefulWidget] that is in charge of attaching a [Game] instance into the flutter tree
+///
+class GameWidget<G extends Game> extends StatefulWidget {
+  /// The game instance in wich this widget will render
+  final G game;
+
+  /// The text direction to be used in text elements in a game.
+  final TextDirection textDirection;
+
+  /// Builder to provide a widget tree to be built whilst the [Future] provided
+  /// via [Game.onLoad] is not resolved.
+  final GameLoadingWidgetBuilder loadingBuilder;
+
+  /// Builder to provide a widget tree to be built between the game elements and
+  /// the background color provided via [Game.backgroundColor]
+  final WidgetBuilder backgroundBuilder;
+
+  /// A map to show widgets overlay.
+  ///
+  /// o use it see [GameWidget.withOverlay]
+  final Map<String, WidgetBuilder> overlayBuilderMap;
+
+  /// Renders a [game] in a flutter widget tree.
+  ///
+  /// Ex:
+  /// ```
+  /// ...
+  /// Widget build(BuildContext  context) {
+  ///   return GameWidget(
+  ///     game: MyGameClass(),
+  ///   )
+  /// }
+  /// ...
+  /// ```
+  const GameWidget({
+    Key key,
+    this.game,
+    this.textDirection,
+    this.loadingBuilder,
+    this.backgroundBuilder,
+  })  : overlayBuilderMap = null,
+        super(key: key);
+
+  /// Renders a [game] in a flutter widget tree alongside widgets overlays.
+  ///
+  /// To use overlays, the game subclass has to be mixed with [HasWidgetsOverlay],
+  /// Ex:
+  /// ```
+  /// ...
+  /// class MyGame extends Game with HasWidgetsOverlay {
+  /// ...
+  /// Widget build(BuildContext  context) {
+  ///   return GameWidget(
+  ///     game: MyGame(),
+  ///     overlayBuilderMap: {
+  ///       "PauseMenu": (ctx) {
+  ///         return Text("A pause menu");
+  ///       },
+  ///     },
+  ///   )
+  /// }
+  /// ...
+  /// ```
+  ///
+  /// To learn more, see:
+  /// - [HasWidgetsOverlay]
+  GameWidget.withOverlay({
+    Key key,
+    this.game,
+    this.textDirection,
+    this.loadingBuilder,
+    this.backgroundBuilder,
+    @required this.overlayBuilderMap,
```

Comment:
why not an optional this.overlayBuilderMap, instead of two constructors?

Reply:
When using overlay, we assert if the game has the mixin.

Reply:
good point